benchmarks/htap/lib/monitoring.py

```diff
@@ -64,66 +73,80 @@ def _update_tpch_stats(self, stats, time_now):
-        dt = (self.data['tpcc'][slot]['timestamp'] - self.data['tpcc_previous'][slot]['timestamp']).total_seconds()
-        return 0 if dt <= 0 else int(delta / dt)
-
-    def tpcc_tps(self):
-        return (sum(self.tpcc_tps_for_slot(slot, slot_type) for slot in range(self.num_oltp_slots))
-                for slot_type in ['ok_count', 'timeout_count', 'err_count'])
-
-    def tpcc_totals(self):
-        return (sum(slot[slot_type] for slot in self.data['tpcc'])
-                for slot_type in ['ok_count', 'timeout_count', 'err_count', 'new_order_count'])
-
+    def tpcc_stats_for_query(self, query):
+        return list(filter(lambda x: x['query'] == query, self.data['tpcc']))
+
+    def filter_last_1s(self, stats):
+        ts = stats[-1]['timestamp'] - 1
+        return list(filter(lambda x: x['timestamp'] >= ts, stats))
+
+    def tpcc_tps(self, query):
+        stats = self.tpcc_stats_for_query(query)
+        if not stats:
+            return (0, 0, 0, 0)
+
+        per_sec = defaultdict(int)
+        for stat in stats:
+            bucket = int(stat['timestamp'])
+            per_sec[bucket] += 1
+        tps = list(per_sec.values())
+        if len(tps) <= 2:
+            return (0, 0, 0, 0)
+
+        # first and last second will not be complete so ignore those
+        tps.pop(0)
+        tps.pop(-1)
+        return (tps[-1], min(tps), int(sum(tps)/len(tps)), max(tps))
+
+    def tpcc_latency(self, query):
+        stats = self.tpcc_stats_for_query(query)
+        if not stats:
+            return (0, 0, 0, 0)
+        lat = list(map(lambda x: math.ceil(x['runtime']*1000), stats))
+        lat_1s = list(map(lambda x: math.ceil(x['runtime']*1000), self.filter_last_1s(stats)))
+        return (int(sum(lat_1s)/len(lat_1s)),
+                min(lat),
+                int(sum(lat)/len(lat)),
+                max(lat))
+
+    def tpcc_total(self, query_type):
+        return self.data['tpcc_totals'][query_type]
 
     def tpch_totals(self):
         return (sum(slot[slot_type] for slot in self.data['tpch'])
                 for slot_type in ['ok_count', 'error_count', 'timeout_count'])
 
     def write_tpcc_stats(self):
-        last_60 = self.data['tpcc_tps_history'][-60:]
-        mins = list(min(slot[idx] for slot in last_60) for idx in range(3))
-        maxs = list(max(slot[idx] for slot in last_60) for idx in range(3))
-        avgs = list(int(sum(slot[idx] for slot in last_60)/len(last_60)) for idx in range(3))
-        now = datetime.now()
-        slot_types = ['OK', 'TIMEOUT', 'ERROR'] # same order as tpcc_tps function
-        for idx in range(3):
-            self.csv_tpcc.write(f'{now}, {slot_types[idx]}, {mins[idx]}, {maxs[idx]}, {avgs[idx]}\n')
+        for query_type in QUERY_TYPES:
+            row = [datetime.now(), query_type, self.tpcc_total(query_type), *self.tpcc_tps(query_type), *self.tpcc_latency(query_type)]
+            self.csv_tpcc.write(', '.join(map(str, row)) + "\n")
```

Comment:
what is this `str` here?

Reply:
the function `str` as to make it into string values :)